front: load conflicts last while fetching scenario data

Pivouane opened the pull request:
Closes #14504 

Frontend will now only fetch conflicts once pathfinding process succeeded

emersion commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -117,13 +117,16 @@ const useScenarioData = (scenario: ScenarioWithDetails, infraId: number) => {
   });
 
+  const isReadyToFetchConflicts =
+    allTrainsSimulated && projectionPath?.pathfindingStatus === 'succeeded';
```
The current train might fail pathfinding, but we still want to load conflicts in that case.

More generally, I don't see why conflict loading should be tied to the projection. We want to load conflicts after all trains are simulated, regardless of the projection status.

Pivouane replied:
Looking back at the original issue I'm not sure I fully understood the root cause. 

The issue mentions "without any pathfinding/simulation being in cache" does this mean that `allTrainsSimulated` was becoming `true` before the simulation was actually complete ? Or was the query being triggered too early for another reason we should investigate ?

If not, do we want some sort of delay before loading conflicts ? Or do we need to check for a different condition than `allTrainsSimulated` ?

(If this issue needs someone more familiar with the codebase I completely understand)

emersion replied:
I've finally got some time to investigate. What happens, in sequence:

- Train schedule list is loading. `timetableItems` is undefined, and `allTrainsSimulated` is true. (There is no train being simulated.)
- Infra is still loading. `timetableItems` is populated with the full list, `allTrainsSimulated` is true. (There is no train being simulated.)
- Simulation for the first train schedule is loading. `allTrainsSimulated` is false.

Could we perhaps drop `allTrainsSimulated` and replace it with a `timetableItems && simulatedTrainsById.length === timetableItems.length` check? This would ensure that the train schedule list has been loaded and all of the train schedules have been simulated (regardless of simulation success - `SimulationSummaryResult` has failure variants too).

I think `allTrainsProjected` suffers from the same issue.

emersion replied:
Hm, that check wouldn't be enough: after all trains have been simulated, when a train is edited, we need to avoid fetching conflicts before we grab the new train simulation - but the number of simulated trains is unchanged because we keep the old simulation while the new one is being loaded.

`timetableItems && simulatedTrainsById.length === timetableItems.length && allTrainsSimulated` should work.

`allTrainsSimulated` is not a great name - inverting the boolean and renaming to `isTrainSimulationLoading` would make things clearer I think.

hhirtz replied:
how would it solve https://github.com/OpenRailAssociation/osrd/issues/14668 where the UI runs with both the old conflicts and the new timetableItems when a timetable item is updated?

emersion replied:
Grabbing `currentData` instead of `data` in the `getTimetableByIdConflicts` query should help.

Pivouane replied:
Done !

hhirtz commented on front/src/applications/operationalStudies/hooks/useScenarioData.ts:
```diff
@@ -117,13 +117,16 @@ const useScenarioData = (scenario: ScenarioWithDetails, infraId: number) => {
     isFetching,
   } = osrdEditoastApi.endpoints.getTimetableByIdConflicts.useQuery(
-    allTrainsSimulated ? { id: scenario.timetable_id, infraId: scenario.infra_id } : skipToken
+    isReadyToFetchConflicts ? { id: scenario.timetable_id, infraId: scenario.infra_id } : skipToken
```
This change should also fix this issue: https://github.com/OpenRailAssociation/osrd/issues/14668

but it doesn't actually. #14668 is triggered because `useConflictFilter` runs with the new timetable items and old conflicts, because conflicts aren't always updated when `timetableItems` change.

There is also this issue: https://github.com/OpenRailAssociation/osrd/issues/15297 where the conflicts aren't updated when adding a duplicate (weirdly enough they are updated when deleting a conflicting train schedule)

so i think this `useQuery` call should involve `timetableItem` somehow, or be reworked into `useEffect` or something?? 🤔 because conflicts defenitely do not just depend on scenario ID and infra ID when `timetableItems` is updated by other parts of the frontend. also i don't know how deleting a train schedule manage to update the conflicts? anyway sorry for this wall of text have a nice evening

emersion replied:
In general `useEffect()` should almost never be required, we abuse it a lot in the codebase: https://react.dev/learn/you-might-not-need-an-effect

> conflicts defenitely do not just depend on scenario ID and infra ID when timetableItems is updated by other parts of the frontend

The rtk-query cache tags should handle invalidation an re-fetching: https://redux-toolkit.js.org/rtk-query/usage/automated-refetching

Maybe rtk-query cache tags are messed up?

emersion replied:
Resolving because this thread is unrelated to this PR/issue.

Pivouane committed:
front: fetch conflicts last while loading train simulation

Signed-off-by: Pivouane (Pierre Santamaria) <pierrenumber1@outlook.com>

Pivouane committed:
front: fetch conflicts last while loading train simulation

Signed-off-by: Pivouane (Pierre Santamaria) <pierrenumber1@outlook.com>

SharglutDev commented:
Could you leave a comment to explain why you are closing the PR ? You didn't answer the last comments you had some weeks ago and I see that you are still assign on the issue.

Pivouane commented:
> Could you leave a comment to explain why you are closing the PR ? You didn't answer the last comments you had some weeks ago and I see that you are still assign on the issue.

Was not intentional. Github apparently automatically closes pull requests with 0 commits on them, I rebased the 1547 commits on main and dropped the single one that was there :(

Pivouane committed:
fixup: changing allTrainsSimulated to isTrainSimulationLoading

Signed-off-by: Pivouane (Pierre Santamaria) <pierrenumber1@outlook.com>

SharglutDev commented:
Ok got it. Could you still answer the comments to say if you handled them or not ? :)

Pivouane committed:
fixup: styling

Signed-off-by: Pivouane (Pierre Santamaria) <pierrenumber1@outlook.com>

emersion reviewed (approved):
LGTM and works well for the initial load!

It doesn't seem like this works when updating a train schedule though: conflicts are fetched again immediately instead of waiting. We should open a new issue about this.

<img width="1919" height="312" alt="Image" src="https://github.com/user-attachments/assets/18bbed26-7e03-4876-b8c7-40b0beb3d35d" />

emersion commented:
I've taken the liberty to edit the PR description to add the "Closes" word before the linked issue, so that the issue is automatically closed when this PR is merged.

SharglutDev commented:
> It doesn't seem like this works when updating a train schedule though: conflicts are fetched again immediately instead of waiting. We should open a new issue about this.

Invalidating the tag probably does not take count of what's in the file where the `useQuery` are. It just re-fetches the data back as soon as it get the "info".
But not sure how rtk handles the order of the re-fetched endpoints when many are invalidated together.

emersion commented:
I think it does take into account `useQuery()`, it only refetches if there is an active subscription, ie. a `useQuery()` which isn't skipped.

Pivouane commented:
Created [this issue](https://github.com/OpenRailAssociation/osrd/issues/16347) as a follow up

emersion commented:
@Pivouane, what's the status of this PR?

Pivouane commented:
> @Pivouane, what's the status of this PR?

Waiting for an approve from @SharglutDev or @hhirtz

SharglutDev reviewed (approved):
Lgtm ✅
